proposals/lip-0039.md

```diff
@@ -309,266 +335,345 @@ proof={siblingHashes:[cc956a85, 3c516152, e041e1c0, 6400721e, 3c32b131],
+        self.binary_bitmap = binary_expansion(self.bitmap).lstrip("0")
+    
+    @property
+    def height(self) -> int:
+        return len(self.binary_bitmap)
+
+    @property
+    def binary_path(self) -> str:
+        # Convert the key to binary string and take the first self.height digits.
+        return binary_expansion(self.key)[:self.height]
+```
 
-```java
-function generateQueryProof(queryKey):
-    let ancestorHashes be an empty array
-    let siblingHashes be an empty array
-    let queryBitmap be an empty string
-    let currentNode be the tree root node
-    if currentNode is defaultEmptyNode:
-        let response be a data structure with:
-            {key=queryKey,
-             value=empty buffer,
-             binaryBitmap=empty string,
-             siblingHashes=siblingHashes,
-             ancestorHashes=ancestorHashes}
-        return response
-
-    let binaryKey=binaryExpansion(queryKey)
-    let h = 0
-
-    while currentNode is branch:
-        append currentNode.hash to ancestorHashes
-        let d be the digit at position h of binaryKey
-        if d == 0:
-            set currentNode to the left child of currentNode
-            set currentNodeSibling to the right child of currentNode
-        else if d == 1:
-            set currentNode to the right child of currentNode
-            set currentNodeSibling to the left child of currentNode
-        if currentNodeSibling is defaultEmptyNode:
-            prepend 0 to queryBitmap
+```python
+@dataclass
+class QueryProof(Query):
+    ancestor_hashes: list[bytes]
+    sibling_hashes: list[bytes]
+```
+
+```python
+@dataclass
+class Proof:
+    sibling_hashes: list[bytes]
+    queries: list[Query]
+```
+
+We make use of the following utility functions:
+
+```python
+def binary_expansion(key: bytes):
+    '''
+    Convert a key to its binary representation
+    '''
+    return f"{int.from_bytes(key, 'big'):0{8*len(key)}b}"
+```
+
+```python
+def are_sibling_queries(q1: Query, q2: Query) -> bool:
+    '''
+    This function returns true if two queries are sibling,
+    i.e. they point to two sibling nodes, false otherwise.
+    '''
+    if len(q1.binary_bitmap) != len(q2.binary_bitmap):
+        return False
+
+    if q1.binary_key[:q1.height - 1] != q2.binary_key[:q.height - 1]:
+        return False
+
+    return q1.binary_key[q1.height - 1] != q.binary_key[q1.height - 1]
+```
+
+```python
+def insert_query(
+    q: Query, queries: list[Query]
+) -> list[Query]:
+    '''
+    This function inserts the a query back in the query list in the correct positon
+    to keep the list sorted by longest binary bitmap (bigger height), breaking ties by smaller key.
+    This corresponds to a bottom-up left-right ordering of the corresponding leaf nodes.
+    We assume that the input queries list is already sorted.
+    '''
+    if len(queries) == 0:
+        return [q]
+
+    # Run a binary search to efficiently find the insertion index.
```

Comment:
I think a binary search is not too technical and the comment should clearly explain what we do. If we remove the binary search we get a O(N^2) routine (N is the number of queries) which we can easily avoid with the binary search.

Reply:
I think the specifications should choose the right data structures and algorithms to be efficient, but should otherwise aim for simplicity. We should also use libraries for standard algorithms where possible (e.g., https://docs.python.org/3/library/bisect.html).

Note that the running time of `insertQuery` in Python is O(n) as insertion in a list is O(n). So instead of
```
queryProofs = insertQuery(q, queryProofs)
```
you can do
```
queryProofs = sorted(queryProofs + [q], key=lambda q: (-q.height, q.key))
```
with basically same time complexity as sorting with Timsort (https://en.wikipedia.org/wiki/Timsort) in Python should be O(n) in an almost sorted listed (you could still add a comment to use bisection for optimization in the implementation). Alternatively, I would use the bisect library.

Reply:
Okay, let us keep it then. Note the running time of `insertQuery` is O(n) due to insertion being O(n) and also with basic list iteration it would stay O(n) not O(n^2).

Reply:
O(n^2) I meant calling insertQuery (O(n)) for each query (O(n)), but we agree then, resolving